src/freenet/client/async/SplitFileInserterStorage.java

```diff
@@ -1699,34 +1717,29 @@ public BlockInsert chooseBlock() {
         // FIXME this should probably use SimpleBlockChooser and hence use lowest-retry-count from each segment?
         // Less important for inserts than for requests though...
         synchronized(cooldownLock) {
-        synchronized(this) {
-            if(status == Status.FAILED || status == Status.SUCCEEDED || 
-                    status == Status.GENERATING_METADATA || failing != null) return null;
-        }
-        // Generally segments are fairly well balanced, so we can usually pick a random segment 
-        // then a random key from it.
-        int segNo = random.nextInt(segments.length);
-        SplitFileInserterSegmentStorage segment = segments[segNo];
-        BlockInsert ret = segment.chooseBlock();
-        if(ret != null) {
-            noBlocksToSend = false;
-            return ret;
-        }
-        // Looks like we are close to completion ...
-        boolean[] tried = new boolean[segments.length];
-        tried[segNo] = true;
-        for(int count=1; count<segments.length; count++) {
-            while(tried[segNo = random.nextInt(segments.length)]);
-            tried[segNo] = true;
-            segment = segments[segNo];
-            ret = segment.chooseBlock();
-            if(ret != null) {
-                noBlocksToSend = false;
-                return ret;
+            synchronized(this) {
+                if(status == Status.FAILED || status == Status.SUCCEEDED ||
+                        status == Status.GENERATING_METADATA || failing != null) return null;
             }
-        }
-        noBlocksToSend = true;
-        return null;
+            // Generally segments are fairly well balanced, so we can usually pick a random segment
+            // then a random key from it.
+            for (int i = 0; i < segments.length; i++) {
+                // Randomly pick one of the remaining segments, using a partial Fisher–Yates shuffle
+                // of all segments indices in segmentIndexPermutation
+                int j = random.nextInt(segments.length - i) + i;
+                int chosen = segmentIndexPermutation[j];
+                segmentIndexPermutation[j] = segmentIndexPermutation[i];
+                segmentIndexPermutation[i] = chosen;
+
+                BlockInsert ret = segments[chosen].chooseBlock();
+                if(ret != null) {
```

Comment:
The style guide [requires](https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.6.2-horizontal-whitespace) a space between `if` and `(`.